tooling: add new CLI for CVE publication by SRC members

mtardy opened the pull request:
This is in reaction to https://github.com/kubernetes/sig-security/issues/169#issuecomment-3444067871.

In short, the official CVE feed draws its information from the k/k CVE issues. People asked for us to provide a more structured information like OSV (right now the information is mostly targeted at human, even in the RSS and JSON feeds). People have been manually maintaining the OSV format for the CVE but we would like to automate it so the way we found was to provide the SRC person writing the issue with tooling that would automatically generate a JSON blob in the issue that we can parse. The first example of this was this CVE on which Tabitha used the new tool: https://github.com/kubernetes/kubernetes/issues/136680. You can see it generated the usual template plus a little HTML detail thing with "OSV format" at the end.

By using curl to fetch the issue content:
```shell
curl -L \
  -H "Accept: application/vnd.github+json" \
  -H "X-GitHub-Api-Version: 2022-11-28" \
  https://api.github.com/repos/kubernetes/kubernetes/issues/136680
```

You can see that the body finishes with an HTML comment:
```
</details>\n\n<!-- generated by srctl v1.0.0 (2665728667a5-dirty, 2026-01-31T19:18:07Z, go1.25.6) -->\n",
```

For more details on the PR itself, it's basically this very large commit a7d416bfef74 ("new: add SRC tooling to declare new CVEs"), that took me some hours of works on many weeks of back and forth with people trying during the SIG security tooling meeting. Then I decided (because I couldn't find the time to finish it to make it usable and fix the issues) to try to write with the help of an AI, which you can see in all the logically separated commits after that are more easily reviewable. Since the tool is pretty simple and most of the API and design was already there it was pretty good at finishing task, fixing small bug and adding simple features.

Also please don't squash this PR as the commits might be useful for history.

mtardy committed:
new: add SRC tooling to declare new CVEs

Signed-off-by: Mahe Tardy <mahe.tardy@gmail.com>

tabbysable commented:
Per the tooling meeting conversation, let's try:
* pressing enter mean "edit the current step"
* entering the index for a step imply "edit"
* specify versions like this:
```
thing: ([introduced] < fixed|"all")

# Please enter the text for your changes. Empty lines and lines
# starting with '#' will be ignored. An empty text aborts the change.
#
# Affected Versions
#
# - $COMPONENT $VERSION_RANGE_1
# - $COMPONENT $VERSION_RANGE_2 ...
# - ...
#
# Example:
# instead of - kube-apiserver: <= v1.31.11
# do this - kube-apiserver: < v1.31.12
```

mtardy committed:
linters: add linters and fix associated issues

Signed-off-by: Mahe Tardy <mahe.tardy@gmail.com>

mtardy committed:
feat: convert email.tmpl to Go template syntax

Replace placeholder variables with Go template syntax to match
issue.tmpl pattern. Adds conditional rendering for optional fields.

Generated-by: Claude Opus 4.5
Signed-off-by: Mahe Tardy <mahe.tardy@gmail.com>

mtardy committed:
test: add export template tests

Add tests for ToIssue, ToEmail, ToSlack functions to verify
template rendering with various CVEData configurations.

Generated-by: Claude Opus 4.5
Signed-off-by: Mahe Tardy <mahe.tardy@gmail.com>

mtardy committed:
fix: re-open file after editor closes to read content

The scanner was reading from tmpFile which had its file position at the
end after writing the initial content. After the editor modifies the file,
we need to re-open it to read from the beginning.

Generated-by: Claude Opus 4.5
Signed-off-by: Mahe Tardy <mahe.tardy@gmail.com>

mtardy committed:
fix: copy initSteps map to avoid shared reference across instances

New() was returning Internal structs that all shared the same initSteps
map reference. Mutations to one instance's steps would affect all others.
Now copies the map using maps.Copy.

Generated-by: Claude Opus 4.5
Signed-off-by: Mahe Tardy <mahe.tardy@gmail.com>

mtardy committed:
fix: close file in ExportToFile to prevent descriptor leak

ExportToFile was not closing the file after writing, causing a file
descriptor leak. Added defer file.Close().

Generated-by: Claude Opus 4.5
Signed-off-by: Mahe Tardy <mahe.tardy@gmail.com>

mtardy committed:
fix: use Go template syntax in slack.tmpl

Replaced $COMPONENT and $SUMMARY placeholders with proper Go template
syntax using CVEData fields. Removed warning banner as template now works.

Generated-by: Claude Opus 4.5
Signed-off-by: Mahe Tardy <mahe.tardy@gmail.com>

mtardy committed:
fix: correct test cases to match parseAffectedVersions behavior

The implementation enforces that all components must be the same across
all version lines. Fixed test cases to use same component and added
explicit test case for mixed components error.

Generated-by: Claude Opus 4.5
Signed-off-by: Mahe Tardy <mahe.tardy@gmail.com>

mtardy committed:
fix: check Close() return value to satisfy errcheck linter

Both io.go and state/state.go were not checking the return value of
file.Close(). Now properly checks and returns errors from Close().

Generated-by: Claude Opus 4.5
Signed-off-by: Mahe Tardy <mahe.tardy@gmail.com>

mtardy committed:
test: add comprehensive unit tests for improved coverage

Added tests for:
- state/state.go: New, focus navigation, status, JSON export/restore, truncateMiddle
- state/step.go: StepNumber ASCII/Name methods, initSteps validation
- state/data.go: parseCVSS for CVSS 3.0, 3.1, 4.0 and error cases
- io.go: instructions helper, firstAvailableEditor

Coverage improved from 17.3% to 73.8% for state package.

Generated-by: Claude Opus 4.5
Signed-off-by: Mahe Tardy <mahe.tardy@gmail.com>

mtardy committed:
tmpl: fixes in templates on guidance

Signed-off-by: Mahe Tardy <mahe.tardy@gmail.com>

mtardy committed:
refactor: use HTML comments for template instructions

Convert italic markdown instructions to HTML comments so they are
treated as comments rather than rendered content. Keeps backticks
for inline code references.

Generated-by: Claude Opus 4.5
Signed-off-by: Mahe Tardy <mahe.tardy@gmail.com>

mtardy committed:
feat: print placeholder issue template when no CVE argument provided

When invoked without arguments, prints usage info and the placeholder
issue template to help users quickly create placeholder CVE issues.

Generated-by: Claude Opus 4.5
Signed-off-by: Mahe Tardy <mahe.tardy@gmail.com>

mtardy committed:
feat: add OSV format support for CVE data

Add osv.go with OSV struct types matching the official OSV schema v1.6.0.
Includes ToOSV() method to convert CVEData to OSV format with:
- CVSS severity (auto-detects v2/v3/v4)
- Affected packages with version ranges (ecosystem: Kubernetes)
- References and credits

Generated-by: Claude Opus 4.5
Signed-off-by: Mahe Tardy <mahe.tardy@gmail.com>

mtardy committed:
feat: include OSV JSON in issue template

Add collapsible details section at the end of issue.tmpl that displays
the vulnerability data in OSV format for easy machine consumption.

Generated-by: Claude Opus 4.5
Signed-off-by: Mahe Tardy <mahe.tardy@gmail.com>

mtardy committed:
test: add OSV schema validation using official OSSF library

Add TestOSVOfficialSchemaCompatibility that validates our OSV JSON output
can be parsed by the official github.com/ossf/osv-schema library, ensuring
compatibility with the OSV ecosystem.

Generated-by: Claude Opus 4.5
Signed-off-by: Mahe Tardy <mahe.tardy@gmail.com>

mtardy committed:
test: add official OSV library implementation for validation

Add osv_official.go which uses the official OSSF OSV schema library
(github.com/ossf/osv-schema/bindings/go/osvschema) to convert CVEData
to OSV format. This implementation exists solely for testing purposes
to validate that our custom lightweight implementation produces
equivalent JSON output to the official library.

Generated-by: Claude Opus 4.5
Signed-off-by: Mahe Tardy <mahe.tardy@gmail.com>

mtardy committed:
test: add comparison test between custom and official OSV implementations

Add TestOSVCustomVsOfficialIdentical that validates our custom OSV
implementation (osv.go) produces equivalent JSON output to the official
OSSF library implementation (osv_official.go). This ensures our
lightweight custom implementation maintains compatibility with the
official OSV schema.

Generated-by: Claude Opus 4.5
Signed-off-by: Mahe Tardy <mahe.tardy@gmail.com>

mtardy commented:
So I couldn't find time to finish this but wanted some people from the SRC to try it. 

So I gave a try to AI stuff to generate the missing templates, add the OSV format part in the issue, generate some unit tests and fix a few bugs. The core, which is 90% of this was written manually by me in https://github.com/kubernetes/sig-security/pull/171/changes/7f5e7e62927418290efee27361f86b62e8eee988. I reviewed the patch and they make sense but would need some extra scrutiny. I think it's in a state where it could be tried now so maybe we could merge a `v1` of this and see how it goes, if that's useful. To be honest it removes a lot of joy from "writing code" like that but this was stuff I was never going to finish soon enough... It's mostly boilerplate, converting from my API to OSV API, or generating templates, so it should be fine.

mtardy committed:
feat: add version and build info to generated issue template

Add version constant (1.0.0) and build information (git commit, dirty
state) extracted from runtime/debug at startup. This is included as an
HTML comment at the end of generated issues for traceability:
<!-- generated by srctl v1.0.0 (abc123456789, 2026-01-31T16:00:00Z, go1.23.0) -->

The build info uses Go's embedded VCS information (vcs.revision,
vcs.modified, vcs.time) which is automatically populated when building
from a git repository.

Generated-by: Claude Opus 4.5
Signed-off-by: Mahe Tardy <mahe.tardy@gmail.com>

tabbysable commented:
/hold until it's ready

tabbysable commented:
This is looking really good, I've been kicking the tires on it a bit.

Please add a github issue number field to the JSON and populate it into the templates. Please also add a fix lead name into the JSON and populate it into the templates.

Again, thank you!

tabbysable commented:
Also, can you make the tool strip off a tailing ".json" from its argument, if present? That would allow an existing CVE to have work continue on it by using tab-completion in the shell. :-D

tabbysable commented:
Something about the way the tool sanitizes `<` in the affected versions field causes it to disappear when pasting the generated issue text into github.

This causes the issue to appear to say the fixed version is the affected version.

tabbysable commented:
can we get an "export all" feature?

mtardy committed:
feat: add export all option to export issue, email, and slack at once

Add (a)ll option to the export menu that generates all three output
formats (issue, email, slack) in a single operation. Refactored the
export logic to use a map-based approach, reducing code duplication
and making it easier to add new export formats in the future.

Generated-by: Claude Opus 4.5
Signed-off-by: Mahe Tardy <mahe.tardy@gmail.com>

mtardy committed:
feat: strip .json suffix from argument for shell tab-completion

Allow users to tab-complete existing CVE-YYYY-NNNNN.json files in the
shell and have the tool automatically strip the .json suffix. This
makes it easier to resume work on an existing CVE.

Generated-by: Claude Opus 4.5
Signed-off-by: Mahe Tardy <mahe.tardy@gmail.com>

mtardy committed:
refactor: extend step numbering to support 0-9 and a-b (12 steps)

Change the step numbering system from 0-9 only (10 steps max) to 0-9
plus a-b (12 steps max). This prepares for adding additional metadata
fields like GitHub issue number and fix lead name.

Key changes:
- Update ASCII() method to return 'a'-'b' for steps 10-11
- Add StepNumberFromASCII() helper for case-insensitive parsing
- Handle step keys before the switch to keep idiomatic Go style
- Update compile-time check from 10 to 12 steps

Generated-by: Claude Opus 4.5
Signed-off-by: Mahe Tardy <mahe.tardy@gmail.com>

mtardy commented:
> This is looking really good, I've been kicking the tires on it a bit.

For reference, this is the first CVE announcements on which it's been used so far:
- https://github.com/kubernetes/kubernetes/issues/136680
- https://github.com/kubernetes/kubernetes/issues/136677
- https://github.com/kubernetes/kubernetes/issues/136678
- https://github.com/kubernetes/kubernetes/issues/136679
- https://github.com/kubernetes/kubernetes/issues/136789

mtardy commented:
> Please add a github issue number field to the JSON and populate it into the templates. Please also add a fix lead name into the JSON and populate it into the templates.

Hey @tabbysable  thanks for the feedback on the srctl thing. So for the issue number I can understand where to put it ~~but for the fix lead it's a bit unclear as I don't see any example of that in https://github.com/kubernetes/kubernetes/issues/136680. Where would you inject that in the templates?~~ (see https://github.com/kubernetes/sig-security/pull/171#issuecomment-3849334203)

mtardy committed:
feat: add GitHub issue URL field (step 'a')

Add a new step 'a' to capture the GitHub issue URL for a CVE. The URL
is validated to match the format https://github.com/ORG/REPO/issues/NUM
(also accepts http:// and trailing slash for flexibility).

The parsed URL is used in templates:
- Slack: appends 'Please see <URL> for more details' when set
- Email: adds 'See the GitHub issue for more details: <URL>' after
  Additional Details section

Includes unit tests covering valid URLs and error cases.

Generated-by: Claude Opus 4.5
Signed-off-by: Mahe Tardy <mahe.tardy@gmail.com>

mtardy committed:
feat: add fix lead field (step 'b') for email signature

Add a new step 'b' to capture the fix lead name, used to sign the email
template. The field supports lazy pre-population via a PrePopulate
function in the Step struct.

Key changes:
- Add PrePopulate func() string field to Step struct
- Fix lead uses git config user.name as draft value in editor
- PrePopulate value only appears in editor temp file, not stored
  unless user saves - exiting without saving keeps the field empty
- Email template signature line only appears when fix lead is set

Generated-by: Claude Opus 4.5
Signed-off-by: Mahe Tardy <mahe.tardy@gmail.com>

mtardy commented:
> Something about the way the tool sanitizes `<` in the affected versions field causes it to disappear when pasting the generated issue text into github.
> 
> This causes the issue to appear to say the fixed version is the affected version.

So I think I fixed everything you asked except for this, I interpreted the fix lead as the person from the SRC writing the announcements.

mtardy commented:
> > Something about the way the tool sanitizes `<` in the affected versions field causes it to disappear when pasting the generated issue text into github.
> > This causes the issue to appear to say the fixed version is the affected version.
> 
> So I think I fixed everything you asked except for this, I interpreted the fix lead as the person from the SRC writing the announcements.

Ah it was a bug in the template, I fixed it. Would love to merge the first version of this now that we have proved it was usable!

mtardy committed:
version: bump patch version number

The tool has been used in beta with its first CVE
https://github.com/kubernetes/kubernetes/issues/136680 and has been
proven useful. We found a few issues however that will be fixed by the
following commits.

The version change will allow to see that a new version has been used in
the issue output.

Signed-off-by: Mahe Tardy <mahe.tardy@gmail.com>

mtardy committed:
feat: add CVE.org URLs to OSV references

Add ADVISORY references to the OSV format output, CVE.org URL:
https://www.cve.org/cverecord?id=CVE-XXXX-XXXXX

This provides standard reference links for vulnerability tracking.

Generated-by: Claude Opus 4.5
Signed-off-by: Mahe Tardy <mahe.tardy@gmail.com>

tabbysable commented:
This is demonstrably good enough for a first version, given that it was used in practice and found helpful.

Additional improvements can be made in a normal PR process.

Thank you so much, again. This was so helpful.

/lgtm

PushkarJ commented:
+1 to @tabbysable 

/lgtm

mtardy commented:
> This is demonstrably good enough for a first version, given that it was used in practice and found helpful.
> 
> Additional improvements can be made in a normal PR process.

Yeah we might have more tuning to come but hopefully this can land now!

> 
> Thank you so much, again. This was so helpful.
> 
> /lgtm

> +1 to @tabbysable
> 
> /lgtm

Thanks, I'm glad it helped, it was fun to write.

mtardy commented:
/unhold
